### PR TITLE
test: update Header test to handle more languages 

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
https://github.com/freeCodeCamp/freeCodeCamp/issues/42890 should have been caught in CI, so hopefully this PR will fail the checks for Italian until we fix the underlying issue.

### CODE REVIEW

**Comment by ShaunSHamilton** [comment]
#42859 Has not been merged yet...

**Comment by ojeytonwilliams** [comment]
Yeah, true, this was a bit early.  I'll take Portuguese out for now.

File: .github/workflows/node.js-tests.yml [hunk 1]
```diff
@@ -130,7 +130,7 @@ jobs:
     strategy:
       matrix:
         node-version: [14.x]
-        locale: [chinese, espanol]
+        locale: [chinese, chinese-traditional, espanol, italian, portuguese]
```
Comment (ojeytonwilliams) [inline .github/workflows/node.js-tests.yml:133]:
```suggestion
        locale: [chinese, chinese-traditional, espanol, italian]
```

**Review by raisedadead — COMMENTED** [review]
I think it was intentional that we have only Chinese & Espanol in the language configs. 

The idea is to cover one of each variant (root, subdomain and subpath) - which we already do. Bugs creeping past this should be OK to fix ad-hoc, maintaining a good balance between the no. of CI checks we need to monitor.

The current coverage of tests + the variants clears the critical path for all the variants of client apps.

**Comment by ojeytonwilliams** [comment]
Yeah, I know may not be possible, but if we can afford to run it for each language I'd like to.  For instance, we wouldn't know that two Italian challenges have the same title without running these tests.

**Comment by raisedadead** [comment]
> For instance, we wouldn't know that two Italian challenges have the same title without running these tests.

Sure. Go ahead and create/update workflows that are a part of the entry point of the issues if you prefer or want. In this case, the entry point would be the Crowdin (both automated, manual) runs for instance.

Expanding the scope of the current CI (unit testing suite) is not a good idea/workaround. A key rule of thumb for testing paradigms is to isolate responsibilities and have little to no overlap between things.

File: .github/workflows/node.js-tests.yml [hunk 2]
```diff
@@ -130,7 +130,7 @@ jobs:
     strategy:
       matrix:
         node-version: [14.x]
-        locale: [chinese, espanol]
+        locale: [chinese, chinese-traditional, espanol, italian]
```
Comment (raisedadead) [inline .github/workflows/node.js-tests.yml:133]: To be reverted, the rest all looks good.

**Review by raisedadead — DISMISSED** [review]

**Comment by ojeytonwilliams** [comment]
> A key rule of thumb for testing paradigms is to isolate responsibilities and have little to no overlap between testing things.

I'm not sure I entirely understand that. I thought the rule of thumb was that applying heuristics was a bad idea, since we can't know exactly what the impact of a given change would be.  Instead we should (if possible) run every test for each change.

Case in point, I wouldn't have guessed that `strip-comments` would only affect the Italian, because they happen to have a `if` inside a comment.  End result, we have to rely on user reports when we could have caught it in CI.

However, if the point is just that our tests take too long to run each one on every single PR, then I understand.

**Comment by raisedadead** [comment]
>  I thought the rule of thumb was that applying heuristics was a bad idea, since we can't know exactly what the impact of a given change would be.

and

> However, if the point is just that our tests take too long to run each one on every single PR, then I understand.

That is not the intention here. Not sure if you are assuming it as such.

The idea is the CI tests should have coverage of all (practically most) features. The current CI runs challenge tests on different variations of the client application that we can have. There is no second-guessing here. The variations are the app hosted at:

- the root `/`
- at a subpath like `/espanol` 
- a subdomain `chinese.` 
 
If a test fails for one variation, it should fail for its equivalent (a logical bug on `/italian` should be caught on `/espanol`). Each variation here is a functional unit. You test each functional unit to be solid as best as you can, covering all use cases. Meaning if a test is broken on a subpath, it is logical to assume that the tests will break for all subpath languages. There is logically no difference among these in terms of implementation. That's one part and is covered by the CI tests scope.

Let me reiterate https://github.com/freeCodeCamp/freeCodeCamp/pull/42894#pullrequestreview-708228915 : *The current CI tests scope is to prevent catastrophic errors, like a complete build failure, preventing us from deploying application*

The other part, as I understand, there could be bugs due to content.

In this case, you are correct in saying that you see a use case where it happened to break Italian for a specific scenario. The scenario (here is an `if` statement in a comment) is not related to the Italian variant of the app but "the content" that created the Italian variant of the app. 

To put it another way, you would probably see similar unforeseen use cases being introduced in Portuguese or Swahili (i.e. languages which we don't even pull yet).

The point being challenge tests are a weak measure of guessing what will break. 

**Relying on ad-hoc fixes for such use-cases** is OK, as I said before or writing tests closer to the entry point (Crowdin workflows pulling the changes - separate discussion).

This is not heuristics but containing/scoping what you want to cover.

That way, you can pin the source of the error closest to where it is happening. The fix could be in content or fixing the platform code to accommodate the new use-case in content (For example, either changing the comment or updating the lib to handle the comment, etc.).

**Comment by ojeytonwilliams** [comment]
I'm not 100% sure where we disagree here, since I think we both are saying that having a test fail as early as possible is ideal.  

> The fix could be in content or fixing the platform code to accommodate the new use-case in content (For example, either changing the comment or updating the lib to handle the comment, etc.).

Yep, agreed.  That's why I think the ideal case would be to run all the tests both on changes to the content and to the libraries since either could cause a bug.

Ironically the tests wouldn't have saved us here, because the solution doesn't include the comments :man_shrugging:.  However, that's just an argument for writing better tests.

**Comment by raisedadead** [comment]
In simple terms we are disagreeing on this (I think):

The challenge tests suites original intent (from their inception) aka "scope" was to ensure the seed + solutions pass the tests in the browser. 

But, we are stretching them to be used for any and every validation we can instead of innovaying new test suites for more scope, such as content parity validations in i18n, etc.

**Comment by ojeytonwilliams** [comment]
Yep, that makes sense.  I think we broadly agree, though not 100%.

Anyways, we're not even doing the tests properly - I've only just realised that the solutions we're testing have the original English comments in them.  Oops.

**Comment by raisedadead** [comment]
Sure, do you want to look into the CI issue that is broken for some reason: https://github.com/freeCodeCamp/freeCodeCamp/pull/42894/checks?check_run_id=3086072508#step:7:1336

**Review by raisedadead — APPROVED** [review]

**Review by raisedadead — APPROVED** [review]
Just remembered we need to add temporary redirects to the Italian news and forum because they do not exist. They should be temporarily redirected to English for now

Interested in making a PR for NGINX?

**Comment by ojeytonwilliams** [comment]
Sure, I can put something together.

**Comment by nhcarrigan** [comment]
This doesn't change anything in our workflows, right? Just allows for local testing, and the option to expand our CI?

**Comment by ojeytonwilliams** [comment]
Exactly right, Nick.

**Review by naomi-lgbt — APPROVED** [review]
🚢 it.

Let's remember to add Portuguese when that goes live.

File: .github/workflows/node.js-tests.yml (hunk 2 above)
Comment (nhcarrigan) [reply .github/workflows/node.js-tests.yml]: Resolving as the change was reverted.
